add bitset.size() and version to index file header

staltz opened the pull request:
For issues #8 and #65, this adds two new fields to the file header: `version` and `bitsetSize`. Note it's a breaking change for any system currently deploying jitdb indexes.

So far there is no `bitset.size()` optimization (that old idea of not storing indexes that are too small), but if we ever want to implement that idea, this new header would make it easier.

arj03 commented on files.js:
```diff
@@ -41,21 +57,33 @@ function loadTypedArrayFile(filename, Type, cb) {
+  loadTypedArrayFile(
+    filename,
+    Uint32Array,
+    (err, { version, seq, count, tarr }) => {
```
you are missing bitsetSize here ;)

staltz replied:
It's not necessary, right? bitsetSize is only relevant in `loadBitsetFile`, and actually it would only be necessary once we start using that field, for that idea #8

arj03 commented on index.js:
```diff
@@ -460,18 +475,20 @@ module.exports = function (log, indexesPath) {
       )
     } else {
-      loadBitsetFile(index.filepath, (err, { seq, bitset }) => {
+      loadBitsetFile(index.filepath, (err, { version, seq, bitset }) => {
```
and count should be here as well

staltz replied:
What's the reason? `count` was already available before this PR, and `count` was not used in this line of code

arj03 commented:
Had a crash when running example with this PR

staltz commented:
What was the crash?

arj03 commented:
Somewhere in slowEqual. There is something there if you specify a path it can't find I think. This is unrelated to this PR, but something that would be nice to fix.

arj03 commented:
To reproduce the problem comment out the first example in example.js, the one with the ship. Then make sure indexes are deleted and run node example. Then second run will fail with:

```
internal/validators.js:92
        throw new ERR_OUT_OF_RANGE(name, 'an integer', value);
        ^

RangeError [ERR_OUT_OF_RANGE]: The value of "sourceStart" is out of range. It must be an integer. Received NaN
    at Buffer.compare (buffer.js:877:5)
    at Object.compareString (/home/arj/dev/jitdb/node_modules/bipf/index.js:326:17)
    at checkEqual (/home/arj/dev/jitdb/index.js:246:12)
    at updateIndexValue (/home/arj/dev/jitdb/index.js:302:32)
    at Object.write (/home/arj/dev/jitdb/index.js:362:16)
    at Stream._writeToSink (/home/arj/dev/jitdb/node_modules/async-flumelog/stream.js:57:32)
    at Stream._handleBlock (/home/arj/dev/jitdb/node_modules/async-flumelog/stream.js:91:12)
    at /home/arj/dev/jitdb/node_modules/async-flumelog/stream.js:133:14
    at Request._callback (/home/arj/dev/jitdb/node_modules/async-flumelog/index.js:101:9)
    at Request.callback (/home/arj/dev/jitdb/node_modules/random-access-storage/index.js:173:8) {
  code: 'ERR_OUT_OF_RANGE'
}
```

staltz commented:
Updated this PR to not save bitsetSize, but note that this PR now **fails** tests.

There's one big change I made, which was the header fields `version`, `seq`, and `count` are now all `UInt32LE`. They used to be `Int32LE`. The tests are failing because some indexes are being saved with `seq=-1`. On the other hand, we want to **not** empty indexes! See #71.

So that's what I'm gonna do in this PR, tackle also #71 in a new commit soon. The reason why I changed to unsigned 32: we don't need negative numbers for these fields, and we gain a lot more range with unsigned. It seems we will need that larger range especially for `seq` (which is "size of the log file this index corresponds to"), and `Int32LE` would support only log files up to 2 GB big.

staltz commented:
Ready for review

arj03 commented on files.js:
```diff
@@ -4,17 +4,38 @@ const TypedFastBitSet = require('typedfastbitset')
+ * ## File format for tarr files
+ *
+ * Each header field is 4 bytes in size.
+ *
```
nice

arj03 commented:
I tried this branch using the recipe [here](https://github.com/ssb-ngi-pointer/jitdb/pull/69#issuecomment-745107067) and still have the same problem. Its really wierd, the diff looks good. Have to double check it

arj03 commented:
Ah, version and seq is switched for some reason

staltz commented:
I'll take a look at that problem, I'll try to fix it in this PR

arj03 commented:
In save it seems:

```
jitdb saving core index: offset +1ms
saving version 1
saving seq 983320378
  jitdb saving core index: timestamp +2ms
saving version 1
saving seq 983320378
  jitdb saving core index: sequence +2ms
saving version 1
saving seq 983320378
  jitdb saving index: value_content_type_vote +3ms
saving version 983320378
saving seq 1
  jitdb saving index: value_content_vote_expression_u26F5 +2ms
saving version 983320378
saving seq 1
  jitdb saving index: value_author_@+UMKhpbzXAII+2x2F7ZlsgkJwIsxdfeFi36Z5Rk1gCfY0=x2Eed25519 +2ms
saving version 983320378
saving seq 1
```

arj03 commented on files.js:
```diff
@@ -41,21 +64,25 @@ function loadTypedArrayFile(filename, Type, cb) {
+function saveBitsetFile(filename, version, seq, bitset, cb) {
   bitset.trim()
-  saveTypedArrayFile(filename, seq, bitset.count, bitset.words, cb)
+  saveTypedArrayFile(filename, seq, version, bitset.count, bitset.words, cb)
```
here is the problem :)

staltz replied:
aaaah

staltz replied:
weird that tests don't break, I'll try adding a failing test first

staltz commented:
Added test that fails because of swapped version and seq.

arj03 commented:
Working now :)
